### ISSUE TITLE
Simulator usage - train and test split for target encoded features to avoid leakage

### ISSUE THREAD

**Issue by mrStasSmirnoff** [fidelity/mabwiser#45]
Hello again dear authors,

Firstly I would like to say that I purchased your paper and it helped quite a lot to gain the "context" of the package and motivated me to go deeper. Now going deeper I was thinking to use the Simulator class to find the optimal model and HP, however, since I am working on contextual bandit there a lot of feature encoding is involved. For example, I am utilizing target encoding on my categorical features, for that, I split the data into a train-test and train my encoder on the train set, then apply this encoder on the test set (targets from the test part are not involved in the encoder training process to avoid leakage). Now, looking at the Simulator class I can only see that the whole dataset can be fed, and then inside the package, the train-test split is going to happen. However, in this case, I would need to target encode my features based on the whole dataset but it would create a leakage since they suppose to be an unknown target on the test set that will be used by the encoder...Therefore my question would be if you could give me a hint on how I can overcome this problem? I hope my explanation is not too messy and again I will be helpful with any kind of advice on that matter.

**Comment by bkleyn** [comment]
Dear mrStasSmirnoff,
 
Using appropriate train and test datasets that avoid information leakage is indeed an important consideration. Since the Simulator is used to evaluate a range of bandit algorithms and parameter configurations one option is to only use your “train” split for this purpose. You can think of the test data used by the Simulator as an additional validation set. Once you’ve identified a specific bandit configuration you can then evaluate this separately on your test set. This should allow you to perform the necessary preprocessing outside of the Simulator class, while maintaining a robust train/test protocol.
